### PR TITLE
More metadata writer API clean-up and testing

### PR DESCRIPTION
**Pull request by tmat**
Addresses https://github.com/dotnet/corefx/issues/7697

### CODE REVIEW

**Comment by tmat** [comment]
@nguerrera

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobWriter.cs [hunk 1]
```diff
@@ -348,18 +348,18 @@ public void WriteDateTime(DateTime value)
...
         {
-            Debug.Assert(size == 2 || size == 4);
+            // This code is a very hot path, hence we don't check if the reference actually fits 2B.
 
-            if (size == 2)
+            if (isSmall)
             {
                 Debug.Assert((ushort)reference == reference);
-                WriteUInt16((ushort)reference);
+                WriteUInt16(unchecked((ushort)reference));
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobWriter.cs]: Why `unchecked` given the above assert? I assume we won't turn on the perf hit of checked arithmetic in release, and to me `unchecked` implies overflow is by design where as overflow here would be a bug in the caller, no?

**Comment by nguerrera** [comment]
>                     break;

I suggest a comment here on why it is no-op.

---
Refers to: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataBuilder.Tables.cs:203 in 2eba093. [](commit_id = 2eba0931e79d0db2b7fe18e1602572216ef13d3e, deletion_comment = False)

**Comment by nguerrera** [comment]
>     // TODO: argument checking

Is this done?

---
Refers to: src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobWriter.cs:13 in 2eba093. [](commit_id = 2eba0931e79d0db2b7fe18e1602572216ef13d3e, deletion_comment = False)

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobWriter.cs [hunk 2]
```diff
@@ -348,18 +348,18 @@ public void WriteDateTime(DateTime value)
         /// <remarks>
         /// References may be small (2B) or large (4B).
         /// </remarks>
-        public void WriteReference(uint reference, int size)
+        public void WriteReference(int reference, bool isSmall)
```
Comment (nguerrera) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobWriter.cs]: I would put the large or small info somewhere more prominent than remarks, parameter doc for isSmall is probably better.

**Comment by nguerrera** [comment]
>                     break;

Add comment about why setting capacity of TableIndex.Assembly(.*)(OS|Processor) or TableIndex.(.*)Ptr is no-op.

---
Refers to: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataBuilder.Tables.cs:203 in 2eba093. [](commit_id = 2eba0931e79d0db2b7fe18e1602572216ef13d3e, deletion_comment = False)

**Comment by nguerrera** [comment]
LGTM

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobWriter.cs (hunk 1 above)
Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobWriter.cs]: OK, I'll remove it.

**Comment by tmat** [comment]
Re:  "TODO: argument checking" -- nope, there is still more work.

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobWriter.cs (hunk 2 above)
Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobWriter.cs]: Done.

**Comment by tmat** [comment]
retest OSX please

**Comment by tmat** [comment]
retest mac please

**Pull request merged**